### ISSUE TITLE
TypeError: Cannot read property 'split' of undefined

### ISSUE THREAD

**Issue by figozhu** [facundoolano/google-play-scraper#192]
version: 5.0.0

invode param:
```javascript
gPlay.list({
        category : gPlay.category.GAME,
        collection : gPlay.collection.NEW_FREE,
        lang : 'zh',
        country : 'CN',
        fullDetail : true,
        start : 0,
        num : 100,
    }).then(function (appList) {
        //console.log(appList);
        cb(null, appList);
    }).catch(function (err) {
        console.error("try to spider googleplay app list failed, err=", err);
        cb(err);
    });

```

log:
```
at parseFields (/googleplay/node_modules/google-play-scraper/lib/app.js:48:64)
    at tryCatcher (/googleplay/node_modules/bluebird/js/release/util.js:16:23)
    at Promise._settlePromiseFromHandler (/googleplay/node_modules/bluebird/js/release/promise.js:512:31)
    at Promise._settlePromise (/googleplay/node_modules/bluebird/js/release/promise.js:569:18)
    at Promise._settlePromise0 (/googleplay/node_modules/bluebird/js/release/promise.js:614:10)
    at Promise._settlePromises (/googleplay/node_modules/bluebird/js/release/promise.js:693:18)
    at Async._drainQueue (/googleplay/node_modules/bluebird/js/release/async.js:133:16)
    at Async._drainQueues (/googleplay/node_modules/bluebird/js/release/async.js:143:10)
    at Immediate.Async.drainQueues (/googleplay/node_modules/bluebird/js/release/async.js:17:14)
    at runCallback (timers.js:794:20)
    at tryOnImmediate (timers.js:752:5)
    at processImmediate [as _immediateCallback] (timers.js:729:5)
```

**Comment by facundoolano** [comment]
Confirmed it fails.

**Comment by Masterxv** [comment]
waiting for fix

**Comment by abhijeet0007** [comment]
I got the same issue 

{
message: "Cannot read property 'split' of undefined"
}

**Comment by timoweiss** [comment]
@facundoolano looks like google is doing some A/B testing. The structure and the css-class naming is different from time to time. Unfortunately, it looks like they are now generating css-class names which results in non-human-readable names (such as `AHFaub`). It could also be the case, that those names change everytime a change to the website is made :(

**Comment by cgwyllie** [comment]
RE @timoweiss it seems that at least some of the fields might be found from attribute selectors using the `itemprop` key, for example `[itemprop="name"]`. Although not everything seems to have such a useful selector available.

**Comment by facundoolano** [comment]
can @cgwyllie or @timoweiss attach an example response html of the new layout?

**Comment by cgwyllie** [comment]
@facundoolano here's one example (Uber Eats listing page): https://gist.github.com/cgwyllie/c2dbb4b148aa2bcd0d4bbc800dc96008

Not sure how much difference there is for other pages etc.

**Comment by yanalavishnu** [comment]
I am facing the same problem.
Can you please fix the problem.
Thanks

**Comment by a2kas** [comment]
Same issue. Anyone has suggestion how to resolve this problem?

**Comment by sunnyrty** [comment]
suggestion how to resolve this problem? my website down for 15days..

**Comment by facundoolano** [comment]
To summarize the current state of this issue: google apparently switched completely to the new layout which uses new class names, most of them not readable (and which could potentially change frequently).

The parsing code in lib/app.js has to be rewritten to reflect these new layout (ideally in a way that would make it easy to update if the class names change again). Alternatively, the hardcoded data in the script tags of the html could be used (see #199), although these are harder to parse and I feel like they could break more easily (although I can't be sure).

I'll be taking a look at updating the code when I get some free time. Those that are in a hurry to get this working again can send a pull request with a fix.

**Comment by facundoolano** [comment]
A bigger issue than the new classes being unreadable, is that different fields share the same classes, thus making it impossible to distinguish which is which (for example, all values in the "Additional information" section share the `.htlgb` class).

**Comment by blakedietz** [comment]
Hey @facundoolano do you have a gitter chat or slack you can hang out in? I would like to help with this current breakage in whatever way possible? Care to chat in the next 10?

**Comment by facundoolano** [comment]
@blakedietz kind of late for me right now, also I'd prefer to keep the conversation async for now.

I think we'll need to explore extracting the data from the javascript arrays in the script tags rather than from html directly.

**Comment by blakedietz** [comment]
Ok, I think that's not a bad way to go right now. The tricky part is identifying which script tag includes the information of note. As it currently stands, I don't see a nice way to identify that.

For those of you that are tuning in at home and want more context, either view source and look for big arrays of data or paste this into your console when looking at the spotify app page:

Url: https://play.google.com/store/apps/details?id=com.spotify.music&showAllReviews=true

```javascript
[...document.querySelectorAll('script[nonce]')].filter(node => node.text.includes('love'))
```

I'm just show that as an example for right now, I know I'm cheating by checking for the word "love". This just shows an example of some of the data.

**Comment by tanqhnguyen** [comment]
We are having the same issue as well, and we have 2 options to work around this

- Use the embedded data as @blakedietz suggested
- Use good old xpath to query the stuff we need

Both have their pros and cons but they are equally fragile as Google seems to keen on protecting their page from web scraping :(

**Comment by nowkin** [comment]
anybody got the right css selector or xpath to get the youtube video url for an app?

**Comment by facundoolano** [comment]
> The tricky part is identifying which script tag includes the information of note. As it currently stands, I don't see a nice way to identify that.

In these cases where we only want to grab stuff from a script tag, I think the best is to just match against the whole html string and then JSON.parse the extracted result (instead of parsing the html). See [this example](https://github.com/facundoolano/app-store-scraper/blob/dc45e7f64b8b09890364235b364fe6de0b482291/lib/similar.js#L21-L28) in the app-store-scraper.

In this case I think we need to match the array between
``` 
AF_initDataCallback({key: 'ds:3', isError:  false , hash: '8', data:function(){return [
```
and
``` 
]
}});</script>
```

(we may need to extract more than one of those, since ds:3 appears to have general app data, and ds:8 has the price.

**Comment by tanqhnguyen** [comment]
So... this is what I have so far

```
let result = {
  title: {
    from: 'ds:3',
    path: '[0][0][0]'
  },
  description: {
    from: 'ds:3',
    path: '[0][10][0][1]'
  },
  logoUrl: {
    from: 'ds:3',
    path: '[0][12][1][3][2]'
  },
  categories: {
    from: 'ds:3',
    path: [
      '[0][12][13][0][0]'
    ]
  },
  ratingValue: {
    from: 'ds:10',
    path: '[0][0][1]' 
  },
  ratingCount: {
    from: 'ds:10',
    path: '[0][2][1]'
  },
  publisherName: {
    from: 'ds:3',
    path: '[0][12][5][1]'
  },
  screenshots: {
    from: 'ds:3',
    path: '[0][12][0]'
  }
};
```
This map contains the basic data to query stuff from the weird google data structure found on a game page

- `path` is whatever passed to `_.get`
- `from` indicates from which "node" we should get the data from

Unfortunately, this is purely manual work to find the correct node and its path :(

And the regex is
```
/<script nonce="[\S]+">AF_initDataCallback\(([\s\S]*?)\);<\/script>/gi
```

I also need to use `vm` to execute the script content matched by the above regex

**Comment by cgwyllie** [comment]
Hey folks, looks like some interesting progress around this.

Just a quick thought, is there any plan to support both the older format and the newer format as a fallback?

From what we've observed it seems not _everything_ on the play store has migrated to this newer markup (yet) so it might be good to try both parsers. Perhaps only newly published apps or app updates now generate the newer format.

**Comment by facundoolano** [comment]
This applies only to parsing the application detail page. AFAIK the rest of the parsers are still workers (or at least tests are passing)

**Comment by facundoolano** [comment]
@tanqhnguyen just a suggestion, it would be neat to parse the scripts into a map object with 'ds:3', 'ds:10' as the keys, and the arrays as the values.

then you could express the field paths like this:

```js
{
 title: ['ds:3', 0, 0, 0],
 description: ['ds:3', 0, 10, 0, 1]
}
```
There are ramda functions to facilitate extracting data from paths like those.

Also, don't use the vm or eval to extract the arrays. You can just remove `AF_initDataCallback(` and replace the strings like `data:function(){return` with `data:` to get a proper json literal.

**Comment by tanqhnguyen** [comment]
That's true and true, perhaps using `vm` in this case is a bit overkill

**Comment by facundoolano** [comment]
@tanqhnguyen I've created #205 following the logic described above. I think that solution is good as long as google doesn't start changing the shape of the data.

I've added the first few fields to the result, now we just need to track each of the remaining ones and add the paths (I'll get to it eventually but any help to move this forward is appreciated)

**Comment by icaroponce** [comment]
Smth I've noticed: It looks like some data may change considering whether you're logged in or not. Only when I'm logged in, for this [app](https://play.google.com/store/apps/details?id=com.facebook.katana&gl=us&hl=en-US) I see `(Beta)` on the title. I still don't know if it affects other data, but maybe it's something else to pay attention here.

**Comment by sunnyrty** [comment]
api/apps/com.vg.moto.bike.robot.transform.flying.pigeon.spy.games

{"message":"Cannot read property 'split' of undefined"}

**Comment by facundoolano** [comment]
This should be fixed in v6.0.0 

Note some fields changed in the result objects, review the README for details.
